cip/CIP2015-09-16-public-type-system-type-annotation.adoc

```diff
@@ -0,0 +1,259 @@
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: pygments
+
+= CIP2015-09-16
+Public Type System & Type Annotation
+
+*Authors:*
+
+Nigel Small <nigel@neotechnology.com> +
+Stefan Plantikow <stefan.plantikow@neotechnology.com>
+
+toc::[]
+
+== Abstract
+
+This CIP specifies the public part of the type system of Cypher and introduces literal syntax for writing down types as well as new syntax for type annotations. The proposed type annotation syntax does not change the Cypher grammar right now but rather is intended as a template to be used by future CIPs. This CIP also informatively discusses future syntax that might use the introduced literal type syntax.
+
+== Introduction
+=== Background and Motivation
+In order to support currently ongoing work for allowing users to define _procedures_ and _functions_ that interact with Cypher, it has become necessary to specify types explicitly, especially as part of procedure and function signatures. It is thus necessary to at least introduce syntax for all valid, supported public types as well as syntax for type ascription.
+
+=== Overview
+For the purposes of this CIP, a type system is a system of classifying expressions according to what kinds of values they may evaluate to during the execution of a Cypher query. It does *not *include a specification of a type inferencing and checking algorithm for Cypher queries.
```

Comment:
Is it only about what types expression results will have? Isn't it also about expectations of value types?